File: Baekjoon/java/1717.java
import java.util.*;
import java.io.*;

public class Main {
	
	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	StringTokenizer st;
	StringBuilder sb = new StringBuilder();
	int n, m;
	int[] root;
	
	int find(int x) {
		if(root[x]==-1) {
			return x;
		}
		return root[x] = find(root[x]);
	}
	
	void union(int cmd, int a, int b) {
		int pa = find(a);
		int pb = find(b);
		if(pa != pb) {
			if(cmd==0) root[pa] = pb;
			else sb.append("NO").append('\n');
		} else {
			if(cmd==1) sb.append("YES").append('\n');
		}
	}
	
	void solve() throws IOException {
		st = new StringTokenizer(br.readLine(), " ");
		n = Integer.parseInt(st.nextToken());
		m = Integer.parseInt(st.nextToken());
		root = new int[n+1];
		Arrays.fill(root, -1);
		for(int i=0; i<m; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int cmd = Integer.parseInt(st.nextToken());
			int a = Integer.parseInt(st.nextToken());
			int b = Integer.parseInt(st.nextToken());
			union(cmd, a, b);
		}
		System.out.println(sb);
	}

	public static void main(String[] args) throws IOException {
		new Main().solve();
	}
}

File: Baekjoon/cpp/1717.cpp
#define _CRT_SECURE_NO_WARNINGS
#include <cstdio>
#include <algorithm>
#include <vector>

int n, m;
int arr[1000010];

int Find(int a) {
	if (arr[a] == a) return a;
	return arr[a] = Find(arr[a]);
}

void Union(int a, int b) {
	int pa = Find(a);
	int pb = Find(b);
	if (pa != pb) arr[pa] = pb;
}

int main() {
	scanf("%d %d", &n, &m);

	for (int i = 0; i <= n; i++)
		arr[i] = i;

	for (int i = 0; i < m; i++) {
		int q, a, b;
		scanf("%d %d %d", &q, &a, &b);
		if (q == 0) {
			Union(a, b);
		}
		else {
			int pa = Find(a);
			int pb = Find(b);
			if (pa != pb) printf("NO\n");
			else printf("YES\n");
		}
	}

	return 0;
}
